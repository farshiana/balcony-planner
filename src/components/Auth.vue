<template>
    <v-container fill-height justify-center fluid>
        <v-col md="4" align="center">
            <v-card :elevation="12">
                <v-toolbar
                    color="primary"
                    dark
                    flat
                >
                    <v-toolbar-title>
                        {{ isRegister ? $t('auth.registerForm') : $t('auth.loginForm') }}
                    </v-toolbar-title>
                </v-toolbar>
                <v-form lazy-validation @submit.prevent="onSubmit">
                    <v-card-text>
                        <v-text-field
                            v-model="username"
                            prepend-icon="mdi-account"
                            :error-messages="usernameErrors"
                            :label="$t('auth.username')"
                            required
                            @blur="$v.username.$touch()"
                        />
                        <v-text-field
                            v-model="password"
                            prepend-icon="mdi-lock"
                            :error-messages="passwordErrors"
                            :label="$t('auth.password')"
                            type="password"
                            :counter="isRegister ? 12 : false"
                            password
                            required
                            @blur="$v.password.$touch()"
                        />
                        <v-text-field
                            v-if="isRegister"
                            v-model="repeatPassword"
                            prepend-icon="mdi-lock"
                            :error-messages="repeatPasswordErrors"
                            :label="$t('auth.repeatPassword')"
                            type="password"
                            password
                            required
                            @blur="$v.repeatPassword.$touch()"
                        />
                    </v-card-text>
                    <v-card-actions>
                        <v-spacer />
                        <v-btn data-testid="submit" color="primary" type="submit" :loading="isSaving" >
                            {{ isRegister ? $t('auth.register') : $t('auth.login') }}
                        </v-btn>
                    </v-card-actions>
                </v-form>
            </v-card>
            <div class="my-3">
                {{ isRegister ? $t('auth.withAccount') : $t('auth.withoutAccount')  }}
                <v-btn data-testid="mode-switch" text @click="onSwitch">
                    {{ isRegister ? $t('auth.login') : $t('auth.register') }}
                </v-btn>
            </div>
        </v-col>
    </v-container>
</template>

<script>
import { mapActions } from 'vuex';
import { validationMixin } from 'vuelidate';
import {
    required, maxLength, minLength, sameAs,
} from 'vuelidate/lib/validators';

export default {
    mixins: [validationMixin],
    validations() {
        const validations = {
            password: { required, minLength: minLength(12) },
            username: { required, maxLength: maxLength(30) },
        };
        if (this.isRegister) {
            validations.repeatPassword = { required, sameAsPassword: sameAs('password') };
        }
        return validations;
    },
    computed: {
        usernameErrors() {
            const errors = [];
            if (!this.$v.username.$dirty) return errors;
            if (!this.$v.username.maxLength) errors.push(this.$t('auth.usernameMaxLength'));
            if (!this.$v.username.required) errors.push(this.$t('auth.usernameRequired'));
            return errors;
        },
        passwordErrors() {
            const errors = [];
            if (!this.$v.password.$dirty) return errors;
            if (!this.$v.password.required) errors.push(this.$t('auth.passwordRequired'));
            if (!this.$v.password.minLength) errors.push(this.$t('auth.passwordMinLength'));
            return errors;
        },
        repeatPasswordErrors() {
            const errors = [];
            if (!this.$v.repeatPassword.$dirty) return errors;
            if (!this.$v.repeatPassword.required) errors.push(this.$t('auth.repeatPasswordRequired'));
            if (!this.$v.repeatPassword.sameAsPassword) errors.push(this.$t('auth.repeatPasswordSameAs'));
            return errors;
        },
    },
    data: () => ({
        isSaving: false,
        isRegister: false,
        username: '',
        password: '',
        repeatPassword: '',
    }),
    methods: {
        ...mapActions('auth', ['login', 'register']),

        async onSubmit() {
            this.$v.$touch();
            if (this.$v.$invalid || this.isSaving) return;

            this.isSaving = true;
            const data = { username: this.username, password: this.password };
            if (this.isRegister) {
                await this.register(data);
            } else {
                await this.login(data);
            }
            this.isSaving = false;
        },
        onSwitch() {
            this.$v.$reset();
            this.isRegister = !this.isRegister;
        },
    },
};
</script>
