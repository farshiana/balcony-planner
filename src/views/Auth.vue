<template>
    <v-container class="fill-height justify-center" fluid>
        <v-col md="4" align="center">
            <v-card :elevation="12">
                <v-toolbar
                    color="primary"
                    dark
                    flat
                >
                    <v-toolbar-title>
                        {{ isRegister ? $t('auth.registerForm') : $t('auth.loginForm') }}
                    </v-toolbar-title>
                </v-toolbar>
                <v-form lazy-validation @submit.prevent="onSubmit">
                    <v-card-text>
                        <v-text-field
                            v-if="isRegister"
                            v-model="name"
                            prepend-icon="mdi-account"
                            :error-messages="nameErrors"
                            :label="$t('name')"
                            required
                            @blur="$v.name.$touch()"
                        />
                        <v-text-field
                            v-model="email"
                            prepend-icon="mdi-email"
                            :error-messages="emailErrors"
                            :label="$t('auth.email')"
                            required
                            @blur="$v.email.$touch()"
                        />
                        <v-text-field
                            v-model="password"
                            prepend-icon="mdi-lock"
                            :error-messages="passwordErrors"
                            :label="$t('auth.password')"
                            type="password"
                            :counter="isRegister ? 12 : false"
                            password
                            required
                            @blur="$v.password.$touch()"
                        />
                        <v-text-field
                            v-if="isRegister"
                            v-model="repeatPassword"
                            prepend-icon="mdi-lock"
                            :error-messages="repeatPasswordErrors"
                            :label="$t('auth.repeatPassword')"
                            type="password"
                            password
                            required
                            @blur="$v.repeatPassword.$touch()"
                        />
                    </v-card-text>
                    <v-card-actions>
                        <v-spacer />
                        <v-btn color="primary" type="submit" :loading="isSaving" >
                            {{ isRegister ? $t('auth.register') : $t('auth.login') }}
                        </v-btn>
                    </v-card-actions>
                </v-form>
            </v-card>
            <div class="my-3">
                {{ isRegister ? $t('auth.withAccount') : $t('auth.withoutAccount')  }}
                <v-btn text @click="onSwitch">
                    {{ isRegister ? $t('auth.login') : $t('auth.register') }}
                </v-btn>
            </div>
        </v-col>
    </v-container>
</template>

<script>
import { mapActions } from 'vuex';
import { validationMixin } from 'vuelidate';
import {
    required, maxLength, email, minLength, sameAs,
} from 'vuelidate/lib/validators';

export default {
    mixins: [validationMixin],
    validations() {
        const validations = {
            email: { required, email },
            password: { required, minLength: minLength(12) },
        };
        if (this.isRegister) {
            validations.name = { required, maxLength: maxLength(30) };
            validations.repeatPassword = { required, sameAsPassword: sameAs('password') };
        }
        return validations;
    },
    computed: {
        nameErrors() {
            const errors = [];
            if (!this.$v.name.$dirty) return errors;
            if (!this.$v.name.maxLength) errors.push(this.$t('nameMaxLength'));
            if (!this.$v.name.required) errors.push(this.$t('nameRequired'));
            return errors;
        },
        emailErrors() {
            const errors = [];
            if (!this.$v.email.$dirty) return errors;
            if (!this.$v.email.required) errors.push(this.$t('auth.emailRequired'));
            if (!this.$v.email.email) errors.push(this.$t('auth.emailValid'));
            return errors;
        },
        passwordErrors() {
            const errors = [];
            if (!this.$v.password.$dirty) return errors;
            if (!this.$v.password.required) errors.push(this.$t('auth.passwordRequired'));
            if (!this.$v.password.minLength) errors.push(this.$t('auth.passwordMinLength'));
            return errors;
        },
        repeatPasswordErrors() {
            const errors = [];
            if (!this.$v.repeatPassword.$dirty) return errors;
            if (!this.$v.repeatPassword.required) errors.push(this.$t('auth.repeatPasswordRequired'));
            if (!this.$v.repeatPassword.sameAsPassword) errors.push(this.$t('auth.repeatPasswordSameAs'));
            return errors;
        },
    },
    data() {
        return {
            isSaving: false,
            isRegister: false,
            name: '',
            email: '',
            password: '',
            repeatPassword: '',
        };
    },
    methods: {
        ...mapActions('auth', ['login', 'register']),

        async onSubmit() {
            this.$v.$touch();
            if (this.$v.$invalid || this.isSaving) return;

            this.isSaving = true;
            const data = { name: this.name, email: this.email, password: this.password };
            if (this.isRegister) {
                await this.register(data);
            } else {
                await this.login(data);
            }
            this.isSaving = false;
        },
        onSwitch() {
            this.$v.$reset();
            this.isRegister = !this.isRegister;
        },
    },
};
</script>
