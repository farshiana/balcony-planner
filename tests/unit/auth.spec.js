import { fireEvent } from '@testing-library/vue';
import Auth from '@/components/Auth.vue';
import render from '../render';

describe('Auth.vue', () => {
    beforeEach(() => {
        fetch.mockResponseOnce(JSON.stringify({ id: 1 }));
    });

    it('shows login form', () => {
        const { container } = render(Auth);
        expect(container).toMatchSnapshot();
    });

    it('logs in', async () => {
        const { getByLabelText, getByTestId } = render(Auth);

        await fireEvent.update(getByLabelText('Username'), 'toto');
        await fireEvent.update(getByLabelText('Password'), 'short_pwd');
        await fireEvent.click(getByTestId('submit'));
        expect(fetch).toHaveBeenCalledTimes(0);

        await fireEvent.update(getByLabelText('Password'), 'longer_password');
        await fireEvent.click(getByTestId('submit'));
        expect(fetch).toHaveBeenCalledTimes(1);
    });

    it('registers', async () => {
        const { getByLabelText, getByTestId } = render(Auth);

        await fireEvent.click(getByTestId('mode-switch'));
        await fireEvent.update(getByLabelText('Username'), 'toto');
        await fireEvent.update(getByLabelText('Password'), 'correct_password');
        await fireEvent.update(getByLabelText('Repeat password'), 'different_password');
        await fireEvent.click(getByTestId('submit'));
        expect(fetch).toHaveBeenCalledTimes(1);

        await fireEvent.update(getByLabelText('Repeat password'), 'correct_password');
        await fireEvent.click(getByTestId('submit'));
        expect(fetch).toHaveBeenCalledTimes(2);
    });
});
