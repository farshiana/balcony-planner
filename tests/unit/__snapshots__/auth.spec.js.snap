// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth.vue shows login form 1`] = `
<div>
  <div
    class="container container--fluid fill-height justify-center"
  >
    <div
      align="center"
      class="col-md-4 col"
    >
      <div
        class="v-card v-sheet theme--light elevation-12"
      >
        <header
          class="v-sheet theme--dark v-toolbar v-toolbar--flat primary"
          style="height: 64px;"
        >
          <div
            class="v-toolbar__content"
            style="height: 64px;"
          >
            <div
              class="v-toolbar__title"
            >
              
                    Login form
                
            </div>
          </div>
        </header>
         
        <form
          class="v-form"
          novalidate="novalidate"
        >
          <div
            class="v-card__text"
          >
            <div
              class="v-input theme--light v-text-field"
            >
              <div
                class="v-input__prepend-outer"
              >
                <div
                  class="v-input__icon v-input__icon--prepend"
                >
                  <i
                    aria-hidden="true"
                    class="v-icon notranslate mdi mdi-account theme--light"
                  />
                </div>
              </div>
              <div
                class="v-input__control"
              >
                <div
                  class="v-input__slot"
                >
                  <div
                    class="v-text-field__slot"
                  >
                    <label
                      class="v-label theme--light"
                      for="input-15"
                      style="left: 0px; position: absolute;"
                    >
                      Username
                    </label>
                    <input
                      id="input-15"
                      required="required"
                      type="text"
                    />
                  </div>
                </div>
                <div
                  class="v-text-field__details"
                >
                  <div
                    class="v-messages theme--light"
                  >
                    <transition-group-stub
                      class="v-messages__wrapper"
                      name="message-transition"
                      tag="div"
                    />
                  </div>
                </div>
              </div>
            </div>
             
            <div
              class="v-input theme--light v-text-field"
            >
              <div
                class="v-input__prepend-outer"
              >
                <div
                  class="v-input__icon v-input__icon--prepend"
                >
                  <i
                    aria-hidden="true"
                    class="v-icon notranslate mdi mdi-lock theme--light"
                  />
                </div>
              </div>
              <div
                class="v-input__control"
              >
                <div
                  class="v-input__slot"
                >
                  <div
                    class="v-text-field__slot"
                  >
                    <label
                      class="v-label theme--light"
                      for="input-19"
                      style="left: 0px; position: absolute;"
                    >
                      Password
                    </label>
                    <input
                      id="input-19"
                      password=""
                      required="required"
                      type="password"
                    />
                  </div>
                </div>
                <div
                  class="v-text-field__details"
                >
                  <div
                    class="v-messages theme--light"
                  >
                    <transition-group-stub
                      class="v-messages__wrapper"
                      name="message-transition"
                      tag="div"
                    />
                  </div>
                </div>
              </div>
            </div>
             
            <!---->
          </div>
           
          <div
            class="v-card__actions"
          >
            <div
              class="spacer"
            />
             
            <button
              class="v-btn v-btn--contained theme--light v-size--default primary"
              data-testid="submit"
              type="submit"
            >
              <span
                class="v-btn__content"
              >
                
                        Login
                    
              </span>
            </button>
          </div>
        </form>
      </div>
       
      <div
        class="my-3"
      >
        
            Don't have an account?
            
        <button
          class="v-btn v-btn--flat v-btn--text theme--light v-size--default"
          data-testid="mode-switch"
          type="button"
        >
          <span
            class="v-btn__content"
          >
            
                Register
            
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
